database2/ffboltdb/db.go

```diff
@@ -0,0 +1,1510 @@
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return false, err
+	}
+
+	return tx.hasBlock(hash), nil
+}
+
+// HasBlocks returns whether or not the blocks with the provided hashes
+// exist in the database.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrTxClosed if the transaction has already been closed
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) HasBlocks(hashes []wire.ShaHash) ([]bool, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return nil, err
+	}
+
+	results := make([]bool, len(hashes))
+	for i := range hashes {
+		results[i] = tx.hasBlock(&hashes[i])
+	}
+
+	return results, nil
+}
+
+// fetchBlockRow fetches the metadata stored in the block index for the provided
+// hash.  It will return ErrBlockNotFound if there is no entry and ErrCorruption
+// if the checksum of the entry doesn't match.
+func (tx *transaction) fetchBlockRow(hash *wire.ShaHash) ([]byte, error) {
+	blockRow := tx.blockIdxBucket.Get(hash[:])
+	if blockRow == nil {
+		str := fmt.Sprintf("block %s does not exist", hash)
+		return nil, makeDbErr(database.ErrBlockNotFound, str, nil)
+	}
+
+	// Ensure the block row checksum matches.  The checksum is at the end.
+	gotChecksum := crc32.Checksum(blockRow[:checksumOffset], castagnoli)
+	wantChecksumBytes := blockRow[checksumOffset : checksumOffset+4]
+	wantChecksum := byteOrder.Uint32(wantChecksumBytes)
+	if gotChecksum != wantChecksum {
+		str := fmt.Sprintf("metadata for block %s does not match "+
+			"the expected checksum - got %d, want %d", hash,
+			gotChecksum, wantChecksum)
+		return nil, makeDbErr(database.ErrCorruption, str, nil)
+	}
+
+	return blockRow, nil
+}
+
+// FetchBlockHeader returns the raw serialized bytes for the block header
+// identified by the given hash.  The raw bytes are in the format returned by
+// Serialize on a wire.BlockHeader.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrBlockNotFound if the request block hash does not exist
+//   - ErrTxClosed if the transaction has already been closed
+//   - ErrCorruption if the database has somehow become corrupted
+//
+// NOTE: The data returned by this function is only valid during a
+// database transaction.  Attempting to access it after a transaction
+// has ended results in undefined behavior.  This constraint prevents
+// additional data copies and allows support for memory-mapped database
+// implementations.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) FetchBlockHeader(hash *wire.ShaHash) ([]byte, error) {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return nil, err
+	}
+
+	// When the block is pending to be written on commit return the bytes
+	// from there.
+	if tx.pendingBlocks != nil {
```

Comment:
I agree with you on the ones that are just checking the map, but I think this one is better to stay because of the call to the function.  I know the function will return nil, but I think it makes the code a little clearer in this instance.